### PR TITLE
chore(compass-aggregation): move preview fetching logic to a separate class COMPASS-6165

### PR DESCRIPTION
**Pull request by gribnoysup**
Following the tech design, we are moving preview fetching management to a separate class outside of redux store so that the fetching logic is separate from the plugin state management

### CODE REVIEW

File: packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/index.tsx [hunk 1]
```diff
@@ -72,7 +71,7 @@ type PipelineHeaderProps = {
   onToggleOptions: () => void;
   isOpenPipelineVisible: boolean;
   isSavedPipelineVisible: boolean;
-  savedPipelines: Pipeline[];
+  savedPipelines: { id: string; name: string }[];
```
Comment (gribnoysup) [inline packages/compass-aggregations/src/components/pipeline-toolbar/pipeline-header/index.tsx:74]: Here and in a few other places: `Pipeline` from `modules/pipeline` is stage state, not saved pipeline type

File: packages/compass-aggregations/src/modules/pipeline-builder/pipeline-preview-manager.ts [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+/**
+ * Ops that must scan the entire results before moving to the
+ * next stage.
+ */
+const FULL_SCAN_OPS = ['$group', '$bucket', '$bucketAuto'];
+
+/**
+ * Stage operators that are required to be the first stage.
+ */
+const REQUIRED_AS_FIRST_STAGE = [
```
Comment (gribnoysup) [inline packages/compass-aggregations/src/modules/pipeline-builder/pipeline-preview-manager.ts:15]: Same here (but 5 years), do we want to change the list?

File: packages/compass-aggregations/src/modules/pipeline-builder/pipeline-preview-manager.ts [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+      FULL_SCAN_OPS.includes(getStageOp(stage))
+    ) {
+      stages.push({ $limit: options.sampleSize });
+    }
+    stages.push(stage);
+  }
+  if (
+    // TODO: super unsure what this is doing, half of these are not even
+    // selectable stage operators in UI
+    !REQUIRED_AS_FIRST_STAGE.includes(getStageOp(stages[stages.length - 1]))
```
Comment (gribnoysup) [inline packages/compass-aggregations/src/modules/pipeline-builder/pipeline-preview-manager.ts:51]: As far as I can see adding this limit after the stage doesn't prevent user from running these. They indeed need to be first stages, but we are not prepending limit here. Do we want to keep this behavior?

File: packages/compass-aggregations/src/components/aggregations/aggregations.jsx [hunk 4]
```diff
@@ -3,7 +3,6 @@ import { connect } from 'react-redux';
 import Pipeline from '../pipeline';
 import { maxTimeMSChanged } from '../../modules/max-time-ms';
 import { collationStringChanged } from '../../modules/collation-string';
-import { toggleSample } from '../../modules/sample';
```
Comment (gribnoysup) [inline packages/compass-aggregations/src/components/aggregations/aggregations.jsx:6]: Drive-by: this slice is not used by anything anymore, missed it in the previous clean-up

File: packages/compass-aggregations/src/modules/pipeline-builder/pipeline-preview-manager.ts [hunk 5]
```diff
@@ -0,0 +1,108 @@
+import type { DataService } from 'mongodb-data-service';
+import type { AggregateOptions, Document } from 'mongodb';
+import { aggregatePipeline } from '../../utils/cancellable-aggregation';
+import { cancellableWait } from '../../utils/cancellable-promise';
+
+/**
+ * Ops that must scan the entire results before moving to the
+ * next stage.
+ */
+const FULL_SCAN_OPS = ['$group', '$bucket', '$bucketAuto'];
```
[the remark this answers is not in the document]
Comment (mcasimir) [reply packages/compass-aggregations/src/modules/pipeline-builder/pipeline-preview-manager.ts]: this one feels something that may be easier to be maintained here: https://github.com/mongodb-js/ace-autocompleter/blob/main/lib/constants/stage-operators.js#L43

especially with `.ts` we could add that as a mandatory `fullScan: boolean`, so we will keep it up to date for all the stages also in future

File: packages/compass-aggregations/src/modules/pipeline-builder/pipeline-preview-manager.ts (hunk 2 above)
Comment (gribnoysup) [reply packages/compass-aggregations/src/modules/pipeline-builder/pipeline-preview-manager.ts:19]: Yeah, moving both of these there makes sense to me, great suggestion, I'll do that!

Comment (gribnoysup) [reply packages/compass-aggregations/src/modules/pipeline-builder/pipeline-preview-manager.ts:19]: Okay, there are also these merge / out stages that we probably also want to derive from these constants, I'll open a separate ticket to clean this up as we are planning to do some work around `ace-autocompleter` package anyway in this project and do all of these in one go

Comment (mcasimir) [reply packages/compass-aggregations/src/modules/pipeline-builder/pipeline-preview-manager.ts:19]: yeah, that would be awesome! there is already something for `$out` and `$merge` there https://github.com/mongodb-js/ace-autocompleter/blob/main/lib/constants/stage-operators.js#L487

File: packages/compass-aggregations/src/components/stage-preview/stage-preview.jsx [hunk 6]
```diff
@@ -111,7 +111,7 @@ class StagePreview extends Component {
     if (this.props.isComplete) {
       if (!this.props.error) {
         return (
-          <Body className={styles['stage-preview-out']}>
+          <Body as="div" className={styles['stage-preview-out']}>
```
Comment (mcasimir) [inline packages/compass-aggregations/src/components/stage-preview/stage-preview.jsx:114]: We should open a PR to make this the default as you suggested somewhere else .. is super easy to forget

**Review by mcasimir — APPROVED** [review]

File: packages/compass-aggregations/src/modules/pipeline-builder/pipeline-preview-manager.ts (hunk 2 above)
Comment (gribnoysup) [reply packages/compass-aggregations/src/modules/pipeline-builder/pipeline-preview-manager.ts:19]: Oh perfect! Opened COMPASS-6181 and this is basically what I'm planning on working on next, so will refactor this pretty soon
